### PR TITLE
Implement OBSPruningModifier

### PR DESCRIPTION
**Pull request by eldarkurtic**
- Implementation of the pruning algorithm described in https://arxiv.org/abs/2203.07259
- Supports experiments on a single device and multiple-devices via DDP
- Optimized for unstructured and block4 version introduced in the paper
- Gradual pruning experiments on a single-device and multiple-devices via DDP are tested for both mask types (unstructured and block4), and the results match the numbers reported in the paper (+- stdev)
- With FP16 mode and the gradual pruning recipe described in the paper (12-layer BERT-base model, 1024 gradients, block-size 50, 12-layer BERT-base teacher), the experiment fits on a single 24GB RTX 3090 card
- the OBSPruningModifier can be used to prune other models too, it's not constrained to BERT models

### CODE REVIEW

**Review by bfineran — CHANGES_REQUESTED** [review]
looks great overall @eldarkurtic. left a few nits to conform to codebase standards. let's make sure to have a plan to test the implementation vs the paper results

**Comment by eldarkurtic** [comment]
@bfineran all code review updates are pushed; the only thing left is to figure out how to fix the pytorch-tests to handle kwargs when calling `initialize`. That should resolve the:
`RuntimeError: grad_sampler dict with data_loader_builder and loss_function must be provided to initialize GradSampler`

**Review by bfineran — APPROVED** [review]
LGTM. thank you @eldarkurtic! added a check for torch version compatibility

**Review by bfineran — APPROVED** [review]

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+        fisher_block_size: int,
+        mask_type: str,
+    ):
+        super().__init__(params)
+        self._num_grads = num_grads
+        self._damp = damp
+        self._fisher_block_size = fisher_block_size
+        self._mask_type = mask_type
+
+        self._finvs = None  # type: List[EmpiricalBlockFisherInverse]
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:351]: Why can't we directly attach the `type` to the variable?

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+
+        # batched f_inv x g: (batch, B, B) x (batch, B) -> (batch, B)
+        finv_g = torch.einsum("bij,bj->bi", self.f_inv, g)
+
+        # scalar denominator for each batch: (batch)
+        alpha = (self.m + torch.einsum("bi,bi->b", g, finv_g)).sqrt().unsqueeze(1)
+        finv_g /= alpha
+
+        # update f_inv with new outer product: (batch, B) x (batch, B) -> (batch, B, B)
+        self.f_inv.baddbmm_(finv_g.unsqueeze(2), finv_g.unsqueeze(1), alpha=-1)
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:560]: Can we change the method name `baddbmm` to something more verbose?

Comment (eldarkurtic) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:560]: `baddbmm` is a method from PyTorch https://pytorch.org/docs/stable/generated/torch.baddbmm.html

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py (hunk 1 above)
Comment (eldarkurtic) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:351]: do you mean like `self._finvs: List[EmpiricalBlockFisherInverse]` or something else?

Comment (bfineran) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py:351]: the `type:` in-line comment is used throughout sparseml. resolving comment

**Review by dbogunowicz — APPROVED** [review]

**Pull request merged**